asset-transfer-private-data/chaincode-go/private_asset_transfer.go

```diff
@@ -0,0 +1,497 @@
+
+	err = ctx.GetStub().DelPrivateData(ownerCollection, assetDeleteInput.ID) // delete the asset
+	if err != nil {
+			return err
+	}
+
+	return nil
+
+}
+
+// DeleteProposal can be used by a user of a rejected transfer properties to
+// remove the propsal from the asset collection and the org specific collection
+func (s *SmartContract) DeleteProposal(ctx contractapi.TransactionContextInterface) error {
+
+	transMap, err := ctx.GetStub().GetTransient()
+		if err != nil {
+			return fmt.Errorf("Error getting transient: " + err.Error())
+		}
+
+	// Asset properties are private, therefore they get passed in transient field
+	transientDeleteJSON, ok := transMap["agree_delete"]
+		if !ok {
+			return fmt.Errorf("asset to delete not found in the transient map")
+		}
+
+	type assetDelete struct {
+			ID string `json:"asset_id"`
+		}
+
+	var assetDeleteInput assetDelete
+	err = json.Unmarshal(transientDeleteJSON, &assetDeleteInput)
+		if err != nil {
+			return fmt.Errorf("failed to unmarshal JSON: %s", err.Error())
+		}
+
+	if len(assetDeleteInput.ID) == 0 {
+		return fmt.Errorf("ID field must be a non-empty string")
+	}
+
+	// delete private details of agreement
+
+	orgCollection, err := getCollectionName(ctx) // get owners collection
+
+	err = ctx.GetStub().DelPrivateData(orgCollection, assetDeleteInput.ID) // delete the asset
+	if err != nil {
+			return err
+	}
+
+	// delete transfer agreement record
+
+	tranferAgreeKey, err := ctx.GetStub().CreateCompositeKey("transferAgreement",[]string{assetDeleteInput.ID}) // create composite key
+		if err != nil {
+			return fmt.Errorf("failed to create composite key: %s", err.Error())
+		}
+
+	err = ctx.GetStub().DelState(tranferAgreeKey) // remove agreement from state
+		if err != nil {
+			return err
+		}
+
+	return nil
+
+}
+
+// Return the ID of submitting client identity. This will be used to identify
+// and verify the owner on chain
+func (s *SmartContract) ReturnID(ctx contractapi.TransactionContextInterface) (string, error) {
+
+	clientID, err := ctx.GetClientIdentity().GetID()
+	if err != nil {
+		return "", fmt.Errorf("failed to get verified OrgID: %s", err.Error())
+	}
+	return clientID, nil
+}
+
+// getCollectionName is an internal helper function to get collection of submitting client identity
+func getCollectionName(ctx contractapi.TransactionContextInterface) (string, error) {
+
+	// Get the MSP ID of submitting client identity
+	clientMSPID, err := ctx.GetClientIdentity().GetMSPID()
+	if err != nil {
+		return "", fmt.Errorf("failed to get verified OrgID: %s", err.Error())
+	}
+
+	// verify that the client is submitting request to peer in their organization
+	peerOrgID, err := shim.GetMSPID()
+	if err != nil {
+		return "", fmt.Errorf("failed getting peer's orgID: %s", err.Error())
+	}
+
+	if clientMSPID != peerOrgID {
+		return "", fmt.Errorf("client from org %s is not authorized to read or write private data on org %s peer", clientMSPID, peerOrgID)
+	}
```

Comment:
The transfer function needs to be called on both the old owner and new owner peers, therefore this check can't be done on the transfer.

Reply:
Following the pattern set by the secured asset example for this function.